osquery/tables/system/darwin/packages.cpp

```diff
@@ -0,0 +1,308 @@
+  }
+
+  if (size_ < ntohl(Header->indexOffset) + sizeof(BOMBlockTable)) {
+    // BOM block table is invalid.
+    return;
+  }
+
+  Table = (BOMBlockTable*)(data_ + ntohl(Header->indexOffset));
+  table_offset_ = ntohl(Header->indexOffset) + sizeof(BOMBlockTable);
+  if (size_ < table_offset_ + ntohl(Table->count) * sizeof(BOMPointer)) {
+    // BOM Pointer size/count is invalid.
+    return;
+  }
+
+  if (size < ntohl(Header->varsOffset) + sizeof(BOMVars)) {
+    // BOM variable table is invalid.
+    return;
+  }
+
+  Vars = (BOMVars*)(data_ + ntohl(Header->varsOffset));
+  vars_offset_ = ntohl(Header->varsOffset) + sizeof(BOMVars);
+  if (size_ < vars_offset_ + ntohl(Vars->count) * sizeof(BOMVar)) {
+    // BOM variables size/count is invalid.
+    return;
+  }
+  valid_ = true;
+}
+
+/// Lookup a BOM pointer and optionally, it's size.
+const char* BOM::getPointer(int index, size_t* length) const {
+  if (ntohl(index) >= ntohl(Table->count)) {
+    // Requested pointer is out of range.
+    return nullptr;
+  }
+
+  const BOMPointer* pointer = Table->blockPointers + ntohl(index);
+  uint32_t addr = ntohl(pointer->address);
+  if (size_ < addr + ntohl(pointer->length)) {
+    // Address value is out of range.
+    return nullptr;
+  }
+
+  if (length != nullptr) {
+    *length = ntohl(pointer->length);
+  }
+  return data_ + addr;
+}
+
+const BOMVar* BOM::getVariable(size_t* offset) const {
+  if (size_ < vars_offset_ + *offset + sizeof(BOMVar)) {
+    // Offset overflows the variable list.
+    *offset = 0;
+    return nullptr;
+  }
+
+  const BOMVar* var = (BOMVar*)((char*)Vars->list + *offset);
+  *offset += sizeof(BOMVar) + var->length;
+  return var;
+}
+
+const BOMPaths* BOM::getPaths(int index) const {
+  size_t paths_size = 0;
+  auto paths = (BOMPaths*)getPointer(index, &paths_size);
+  if (paths == nullptr || paths_size < sizeof(BOMPaths)) {
+    return nullptr;
+  }
+
+  // Check the number of indexes.
+  if (paths_size < ntohs(paths->count) * sizeof(BOMPathIndices)) {
+    return nullptr;
+  }
+  return paths;
+}
+
+void genBOMPaths(const std::string& path,
+                 const BOM& bom,
+                 const BOMPaths* paths,
+                 QueryData& results) {
+  std::map<uint32_t, std::string> filenames;
+  std::map<uint32_t, uint32_t> parents;
+
+  while (paths != nullptr) {
+    for (unsigned j = 0; j < ntohs(paths->count); j++) {
+      uint32_t index0 = paths->indices[j].index0;
+      uint32_t index1 = paths->indices[j].index1;
+
+      auto info1 = (const BOMPathInfo1*)bom.getPointer(index0);
+      if (info1 == nullptr) {
+        // Invalid BOMPathInfo1 structure.
+        return;
+      }
+
+      auto info2 = (const BOMPathInfo2*)bom.getPointer(info1->index);
+      if (info2 == nullptr) {
+        // Invalid BOMPathInfo2 structure.
+        return;
+      }
+
+      // Compute full name using pointer size.
+      size_t file_size;
+      auto file = (const BOMFile*)bom.getPointer(index1, &file_size);
```

Comment:
This will crash if getPointer failed or the size is less than the structure size.